src/Datadog.NativeLoader/cor_profiler.cpp

```diff
@@ -0,0 +1,881 @@
+#include "cor_profiler.h"
+
+#include "logging.h"
+#include "proxy.h"
+
+namespace datadog
+{
+namespace nativeloader
+{
+
+    CorProfiler::CorProfiler(DynamicDispatcher* dispatcher) : m_refCount(0), m_dispatcher(dispatcher)
```

Comment:
Oh. GH had hidden this file from me, because it was so long. I read everything and was confused cos I was missing the actual forwarder and now I found it.

I think, it is possible to make the design both faster and simpler at the cost of being lee generic. Please let me know what you think of the following:

Right now the dispatcher reads the config file and discovers all "destinations" from there. It is capable of forwarding to any number of profilers. This is great, generic and presumably simple but I think there are some pitfalls. Please correct me if I am overlooking something. :)

Performance.

The dispatcher contains a list of pointers to ICorProfilerCallback. Every time, for every invocation of any profiling API (if selected by the mask). So we always pay for the creation of the lambda closure, for the traversal of the loop and for the invocations of each pointer.

Output params.

Some callback functions provide results via parameters, e.g. `JITCachedFunctionSearchStarted(..)`. Each implementation we forward to may not be aware that it is sharing the profiler hook. So it simply sets the output, and the last one wins. But we may actually prefer some kind of logical combination that is specific to the method. E.g., in this case, only use the cached version if all the profilers agree.

Controller params.

Do we know after calling a callback function that may rewrite IL whether it did so, and would it make sense to restrict others from doping so?

In the context of these challenges, consider that in reality we are not trying to forward to some unknown number of generic profiler implementations. Right now we have two. Over the years we may have 3 or 4 or so ICorProfilerCallbackXx implementations, but not a ton.

So, consider modifying the syntax of the config file to have a "Tracer" and a "Continuous profiler" sections that specify the binary files and architectures just like now. But now we have 2 "strongly typed" forwarding destinations.

With this, there is no need for a generic dispatcher. We have exactly 2 pointers to a ICorProfilerCallbackX implementations. Each of them is dynamically initialized just lie you already do, and each of them may be null if no respective destination was specified in the config.
With this approach, each API `FooCallback(..)` on your `CorProfilerCallbackProxy` class just invokes `_pCorProfilerCallback_ContiniousProfiler->FooCallback(..)` and `_pCorProfilerCallback_Tracer->FooCallback(..)` one after the other. No loop, no lambda.

For the few cases where out params or controllers are passes, some custom logic between the invocations may be necessary.

If in the future we add a third possible destination, we add a new custom section into the config file and a few optionally initialized pointer, e.g. `_pCorProfilerCallback_AppSec->FooCallback(..)` etc.

What do you think of this approach?

Reply:
As we talked offline, a modification is in process to remove the generic approach to something tailored for the tracer and the profiler.

Reply:
cool! thanks!!